### ISSUE TITLE
file::url component throws error in 3.6.4

### ISSUE THREAD

**Issue by texnixe** [k-next/kirby#4274]
## Description
When adding the file::url component as described in the ["Virtual pages from images" recipe](https://getkirby.com/docs/cookbook/templating/virtual-pages-image-gallery), the Panel throws a "The JSON response could not be parsed" error .
Additionally, I get a 502 Bad Gateway error.

The example works perfectly in 3.5.8.

## To reproduce
Add the plugin as described in the recipe to a fresh 3.6.4 Starterkit (probably the same in all 3.6.x versions, not tested yet).

## Your setup
**Kirby Version**  
3.6.4
**Console output**  

**Your system (please complete the following information)**
MacOs 12.1 with Firefox 99
Laravel Valet as dev environment

Probably related to the same changes that lead to this issue: https://github.com/getkirby/kirby/issues/4143

**Comment by texnixe** [comment]
Additional information:

When I change `$file->name()` to `$file->filename()` in the component, the error disappears, but the issue is then the same as in #4143, i.e. the file::url component doesn't have any effect anymore.

**Comment by afbora** [comment]
### Issue summary

When you call a `$file` method that doesn't have but have for `asset` (proxy method) in `file::url` component, recursion calling occurs and throws `Fatal error: Allowed memory` exception.

The summary of the issue is like calling `$file->url()` inside the `file::url` component.

### For example

````php
<?php

Kirby::plugin('cookbook/virtual-gallery', [
  'components' => [
    'file::url' => function (Kirby $kirby, $file) {
      return $file->name();
    }
  ]
]);
````

### Follow the process to understand

1. Call `$file->name()` in `file::url` component
2. `Kirby\Cms\File $file` object has not `name` method
3. Check the proxy method: 
https://github.com/getkirby/kirby/blob/3.7.3/src/Cms/File.php#L106-L109
4. `Kirby\Filesystem\File` has `name` method
5. Try to instance for proxy class
6. Needed `root` and `url` props to instance: 
https://github.com/getkirby/kirby/blob/3.7.3/src/Filesystem/IsFile.php#L105
7. `$this->url()` trigger `file::url` component again
https://github.com/getkirby/kirby/blob/3.7.3/src/Cms/File.php#L616
8. See issue

### Possible solution

I've not solution yet 🙂

**Comment by lukasbestle** [comment]
Hm, tricky...

Maybe we should add a `name()` method to `Cms\File`? This would also increase performance as the proxy can be avoided. But I wonder where this will lead us. `name()` is certainly not going to be the only "missing" method in `Cms\File`.

**Comment by distantnative** [comment]
@afbora Could you try if mirroring `Filesystem\File::name()` in `Filesystem\IsFile` works? I think we do something similar already for `::type` to avoid this type of recursion. Great tracking down what's at the core of the issue!

**Comment by afbora** [comment]
@lukasbestle However, the same error will occur when calling any proxy method in the `file::url` component.

@distantnative I've tested and works. But as I said above, you will get the same error when you call another proxy method, not just name. For ex: `$file->base64()`

**Comment by afbora** [comment]
IMO logical solution; the `file::url` component should not be working/running again while the proxy class is being instantiated .

**Comment by lukasbestle** [comment]
I guess the question is which methods are needed for a custom `file::url` component. `name()` certainly is, so I think it makes sense to add it to `Filesystem\IsFile`.

Maybe we could throw an exception when any proxied method is used from `file::url`. The exception will only affect a few people and they can tell us if they need additional methods.

**Comment by distantnative** [comment]
Another attempt could be to not call `$this->url()` in `IsFile::asset()` but instead use `$this->url` - so only pass a URL if it is already set, no fallback to the component again. But I can't oversee at the moment if this could cause other problems.

**Comment by afbora** [comment]
Btw here source of this change, may be give you an idea about that: https://github.com/getkirby/kirby/pull/3285

**Comment by distantnative** [comment]
Based on that line https://github.com/getkirby/kirby/pull/3285/files#diff-3b700032469a1feb9c9a9eab21e85c3b799a16b107327bc5173c1e375e962f06L172 I could imagine not passing a URL to the asset object could work (or passing `$this->url` which could have a URL or be null).

**Comment by afbora** [comment]
I was just testing it now. Just sending the `root` param solves the problem. But as you say, the critical question is, does it break anything else? 🤷‍♂️

**Comment by distantnative** [comment]
That's at least how it was before https://github.com/getkirby/kirby/pull/3285

**Comment by afbora** [comment]
Unfortunately, this breaks the image or file html output: ` <?= $image ?>` or `<?= $file ?>`. This calls the proxy `->html()` method and the images are not loaded because the `url` is empty.

https://github.com/getkirby/kirby/blob/3.7.3/src/Image/Image.php#L138-L141

**Comment by afbora** [comment]
The idea that came to my mind:

What if instead of taking a `url` parameter, the proxy class itself calls the `file::url` component in the `::url()` method? Thus, all but the `->url()` method will work in the `file::url` component. But for this, the proxy class needs to be aware of the source `Kirby\Cms\File` object.

````php
// https://github.com/getkirby/kirby/blob/3.7.3/src/Filesystem/IsFile.php#L112
new File(['root' => $this->root()]);

// https://github.com/getkirby/kirby/blob/3.7.3/src/Filesystem/File.php#L509-L512
public function url(): ?string
{
    return $this->url ??= ($this->kirby()->component('file::url'))($this->kirby(), $this); // $this object will not work :))
}
````

**Comment by lukasbestle** [comment]
A possible limitation could be that it would no longer be possible to override the `url()` for specific files when we introduce file models later. But since we don't support file models at the moment, that wouldn't be an issue now.

We could introduce a new optional prop for `Filesystem\File` to pass the file object (the name for that prop is probably the most difficult part). We could then also use `$file->kirby()` to get a matching `App` object. If no file object is given, the automatic fallback would be skipped.

**Comment by afbora** [comment]
That's what I meant in my last comment. What do you think of the name `model`?

````php
new File([
  'model' => $this, // Kirby\Cms\File
  'root'  => $this->root()
]);
````

**Comment by lukasbestle** [comment]
Sounds great! 👍

**Comment by afbora** [comment]
@lukasbestle I've tried but current tests fail. Do you think my implementation is wrong? Here is the branch: https://github.com/getkirby/kirby/compare/develop...fix/4274-model-for-filesystem

**Comment by lukasbestle** [comment]
@afbora The `setModel()` method was missing.

I added that but now the next issue was that the `FileVersion` tests failed. `FileVersion` also uses the `Filesystem\IsFile` trait, but PHP doesn't support type-hinting traits. With PHP 8.0 we could define both the `Cms\File` and `Cms\FileVersion` classes, but what would happen with other classes that implement `IsFile` later?

I used a manual check with `class_uses()`, but now Psalm doesn't like the manual docblock type-hint `Kirby\Filesystem\IsFile|null` because again `IsFile` is not a class but a trait. :(

**Comment by afbora** [comment]
@lukasbestle Thanks for take care. 

A class that currently uses the `IsFile` trait needs to have a `url()` method. What if we allow `Object` type instead of defining classes and check the `url()` method?

**Comment by lukasbestle** [comment]
Sounds sensible. If it walks like a file and quacks like a file, it's probably a file. 😂

**Comment by afbora** [comment]
Will be fixed in 3.8.0 ✅
